src/uwtools/utils/api.py

```diff
@@ -5,31 +5,30 @@
 import datetime as dt
 import re
 from pathlib import Path
-from typing import Any, Callable, Optional, Union
+from typing import Any, Callable, Optional, TypeVar, Union
 
-from uwtools.config.formats.base import Config
 from uwtools.drivers.driver import DriverT
 from uwtools.drivers.support import graph
 from uwtools.drivers.support import tasks as _tasks
 from uwtools.exceptions import UWError
 
+T = TypeVar("T")
+
 # Public
 
 
-def ensure_data_source(
-    data_source: Optional[Union[dict, Config, Path, str]], stdin_ok: bool
-) -> Any:
+def ensure_data_source(data_source: T, stdin_ok: bool) -> T:
```

Comment:
I think this is new in the `uwtools` codebase. Python's `TypeVar` provides behavior similar to so-called [generics](https://en.wikipedia.org/wiki/Generic_programming) in other languages. This says "I return a value that is of the same type as the `data_source` argument's value". This is a stronger statement than would be made by e.g.
```
def foo(val: Union[bool, int]) -> Union[bool, int]:
```
where, say, it would be fine for `foo()` to return an `int` value when its `val` argument was an `int`.